Sources/SwiftWarningControl/WarningControlRegions.swift

```diff
@@ -0,0 +1,313 @@
+  mutating func insert(_ region: WarningControlRegion) {
+    let newNode = WarningControlRegionNode(region: region)
+    insertIntoSubtree(newNode, parent: rootRegion)
+  }
+
+  /// Insert a region node into the appropriate position in a subtree
+  private func insertIntoSubtree(
+    _ node: WarningControlRegionNode,
+    parent: WarningControlRegionNode
+  ) {
+    precondition(parent.contains(node) || parent.hasSameBoundaries(with: node))
+    // Check if the new region has the same boundaries as the parent
+    if parent.hasSameBoundaries(with: node) {
+      for region in node.regions.values {
+        parent.addRegion(region)
+      }
+      return
+    }
+
+    // Check if this should be a child of one of parent's children
+    if let containingChild = parent.findChildContaining(node) {
+      insertIntoSubtree(node, parent: containingChild)
+      return
+    }
+
+    // Check if this node should absorb any of the parent's children
+    var absorbedChildren: [WarningControlRegionNode] = []
+    for child in parent.children {
+      if node.contains(child) && !node.hasSameBoundaries(with: child) {
+        absorbedChildren.append(child)
+      }
+    }
+
+    // Remove absorbed children from parent and add to new node
+    parent.children.removeAll { child in absorbedChildren.contains { $0 === child } }
+    for absorbed in absorbedChildren {
+      node.addChild(absorbed)
+    }
+
+    // Add as direct child of parent
+    parent.addChild(node)
+  }
+
+  public func debugDescription() -> String {
+    var result = "Warning Group Control Region Tree:\n"
+    func printNode(_ node: WarningControlRegionNode, indent: Int) {
+      let spacing = String(repeating: "  ", count: indent)
+      result += "\(spacing)[\(node.bounds.start), \(node.bounds.end)]"
+      if !node.regions.isEmpty {
+        result += " id(s): \(node.regions.keys.map { $0 }.joined(separator: ", "))\n"
+      } else {
+        result += "\n"
+      }
+      for child in node.children {
+        printNode(child, indent: indent + 1)
+      }
+    }
+
+    printNode(rootRegion, indent: 0)
+    return result
+  }
+}
+
+extension WarningControlRegionTree {
+  /// Determine the warning group behavior at a specified position
+  /// for a given diagnostic group
+  @_spi(ExperimentalLanguageFeatures)
+  public func warningGroupBehavior(
+    at location: AbsolutePosition,
+    for diagnosticGroupIdentifier: String
+  ) -> WarningGroupBehavior? {
+    return rootRegion.innermostContainingRegion(at: location, for: diagnosticGroupIdentifier)?.behavior
+  }
+}
+
+/// A node in the warning control region tree, representing a region and
+/// references to its nested child regions.
+private class WarningControlRegionNode {
+  let bounds: (start: AbsolutePosition, end: AbsolutePosition)
+  var regions: [String: WarningControlRegion] = [:]  // Keyed by diagnostic group identifier
+  var children: [WarningControlRegionNode] = []
+  weak var parent: WarningControlRegionNode?
```

Comment:
Do we need to keep track of the `parent` here? I don’t see where it’s used and if it is, aren’t we already traversing the tree top-down, so that we would know the parent by the traversal order?